src/rgw/rgw_rados.cc

```diff
@@ -8327,7 +8414,21 @@ int RGWRados::check_disk_state(librados::IoCtx io_ctx,
     // encode a suggested removal of that key
     list_state.ver.epoch = io_ctx.get_last_version();
     list_state.ver.pool = io_ctx.get_id();
-    cls_rgw_encode_suggestion(CEPH_RGW_REMOVE, list_state, suggested_updates);
+
+    if (! object.gc_tag.empty()) {
+      cls_rgw_obj_chain chain;
+      for (auto& it : object.gc_objs) {
+        chain.objs.emplace_back(it);
+      }
+      ldout(cct, 20) << "INFO: check_disk_state going to add objects to gc =" << " obj= " << object.key.name << dendl;
+      r = send_chain_to_gc(chain, object.gc_tag, true);
```

Comment:
i worry that the dir_suggest changes could result in an object's tail segments being submitted for gc more than once (ie during the PutObj/DeleteObj, then again in dir_suggest). since the tail objects are reference counted to support efficient copy (see RGWRados::copy_obj()), the actual gc process uses cls_refcount_put() to decrement its refcount and only delete on last ref. if we schedule a given object for gc more than once, we could end up deleting it while another copy still holds a reference

the defer_gc strategy _could_ help to make sure that any previous gc entries in the queue get skipped, except that it's possible that gc already processed the earlier entry and decremented the refcounts

Reply:
see `cls_rc_refcount_put()` in cls/refcount/cls_refcount.cc:
```c++
  if (objr.refs.empty()) {
    return cls_cxx_remove(hctx);
  }
```

Reply:
ok